### ISSUE TITLE
Do not keep python GIL

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#2955]
In a multi-threaded context created from within a Python extension module, like #2954 ,
we would still like to check for interrupts, print, etc 

See also #2442

We probably need to release GIL for any swig calls, and re-acquire for printing etc

**Comment by lfse-slafleur** [comment]
Hi @jgillis !

We are using a solver which uses Casadi as a dependency in the OMOTES project (In case you are interested: https://github.com/Project-OMOTES , we are building simulation and optimization models for the design of heat nets). We were having some issues where some Python threads appear to be blocked/starved from executing as another thread holds the Python GIL. It seems that Casadi is holding the GIL for the duration of some calculations. Specifically, we are using the HiGHs solver.

I have managed to replicate the scenario in a simpler script (with the ipopt solver) where I took the race car example and added a thread that prints the time every second (if it has the chance):
```python
import datetime
import threading
import time


class PrintNowThread:
    def __init__(self):
        self._thread = threading.Thread(target=self.run)
        self._thread.daemon = True

    def start(self):
        self._thread.start()

    def run(self):
        while True:
            print("____________________________________________________________________")
            print(
                "                                                      The time is",
                datetime.datetime.now(),
            )
            print("____________________________________________________________________")
            time.sleep(1)


PrintNowThread().start()
time.sleep(5)

from casadi import *

N = 500  # number of control intervals

opti = Opti()  # Optimization problem

# ---- decision variables ---------
X = opti.variable(2, N + 1)  # state trajectory
pos = X[0, :]
speed = X[1, :]
U = opti.variable(1, N)  # control trajectory (throttle)
T = opti.variable()  # final time

# ---- objective          ---------
opti.minimize(T)  # race in minimal time

# ---- dynamic constraints --------
f = lambda x, u: vertcat(x[1], u - x[1])  # dx/dt = f(x,u)

dt = T / N  # length of a control interval
for k in range(N):  # loop over control intervals
    # Runge-Kutta 4 integration
    k1 = f(X[:, k], U[:, k])
    k2 = f(X[:, k] + dt / 2 * k1, U[:, k])
    k3 = f(X[:, k] + dt / 2 * k2, U[:, k])
    k4 = f(X[:, k] + dt * k3, U[:, k])
    x_next = X[:, k] + dt / 6 * (k1 + 2 * k2 + 2 * k3 + k4)
    opti.subject_to(X[:, k + 1] == x_next)  # close the gaps

# ---- path constraints -----------
limit = lambda pos: 1 - sin(2 * pi * pos) / 2
opti.subject_to(speed <= limit(pos))  # track speed limit
opti.subject_to(opti.bounded(0, U, 1))  # control is limited

# ---- boundary conditions --------
opti.subject_to(pos[0] == 0)  # start at position 0 ...
opti.subject_to(speed[0] == 0)  # ... from stand-still
opti.subject_to(pos[-1] == 1)  # finish line at position 1

# ---- misc. constraints  ----------
opti.subject_to(T >= 0)  # Time must be positive

# ---- initial values for solver ---
opti.set_initial(speed, 1)
opti.set_initial(T, 1)

# ---- solve NLP              ------
opti.solver("ipopt")  # set numerical backend
sol = opti.solve()  # actual solve

time.sleep(3)
```

If I run this, I receive the following output:
```
$ python3 casadi_example.py
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:17.277698
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:18.279040
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:19.873511
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:20.942321
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:21.960633
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:23.002730
____________________________________________________________________

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.11, running with linear solver MUMPS 5.4.1.

____________________________________________________________________
Number of nonzeros in equality constraint Jacobian...:     4503
                                                      The time is 2024-06-26 11:09:24.596544
____________________________________________________________________
Number of nonzeros in inequality constraint Jacobian.:     1503
Number of nonzeros in Lagrangian Hessian.............:     1502

Total number of variables............................:     1503
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:     1003
Total number of inequality constraints...............:     1002
        inequality constraints with only lower bounds:        1
   inequality constraints with lower and upper bounds:      500
        inequality constraints with only upper bounds:      501

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  1.0000000e+00 1.00e+00 4.12e-01  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  1.0176583e+00 9.90e-01 4.11e-01  -1.7 2.19e+00    -  1.01e-02 1.03e-02h  1
   2  1.0185460e+00 9.89e-01 1.25e+02  -1.7 2.87e+00    -  2.53e-02 4.08e-04h  1
   3  1.0219429e+00 9.88e-01 4.97e+03  -1.7 3.11e+00    -  4.79e-02 1.51e-03h  1
   4  1.0312037e+00 9.84e-01 6.17e+04  -1.7 3.10e+00    -  4.66e-02 4.08e-03h  1
   5  1.0325700e+00 9.83e-01 1.44e+07  -1.7 3.08e+00    -  1.25e-01 5.76e-04h  1
   6  1.0341522e+00 9.83e-01 2.19e+08  -1.7 3.08e+00    -  1.01e-02 6.56e-04h  1
   7  1.0343756e+00 9.83e-01 2.23e+10  -1.7 3.08e+00    -  8.82e-03 9.15e-05h  1
   8r 1.0343756e+00 9.83e-01 1.00e+03  -0.0 0.00e+00    -  0.00e+00 4.73e-07R  3
   9r 1.0105133e+00 9.47e-01 9.91e+02  -0.0 1.55e+01    -  9.31e-03 4.32e-02f  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  10r 1.0928585e+00 8.69e-01 8.75e+02  -0.0 2.72e+00    -  1.17e-01 1.16e-01f  1
  11  1.0948877e+00 8.68e-01 6.36e+01  -1.7 2.92e+00    -  4.11e-02 6.96e-04h  1
  12  1.0950361e+00 8.68e-01 4.39e+04  -1.7 3.11e+00    -  3.24e-02 4.77e-05h  1
  13  1.0965508e+00 8.68e-01 3.12e+06  -1.7 3.16e+00    -  3.42e-02 4.79e-04h  1
  14  1.0975136e+00 8.67e-01 4.87e+08  -1.7 3.17e+00    -  4.72e-02 3.03e-04h  1
  15  1.0976066e+00 8.67e-01 7.82e+11  -1.7 3.19e+00    -  4.71e-02 2.92e-05h  1
  16  1.1039555e+00 8.66e-01 7.49e+12  -1.7 3.26e+00    -  1.88e-02 1.95e-03h  1
MUMPS returned INFO(1) = -9 and requires more memory, reallocating.  Attempt 1
  Increasing icntl[13] from 1000 to 2000.
MUMPS returned INFO(1) = -9 and requires more memory, reallocating.  Attempt 2
  Increasing icntl[13] from 2000 to 4000.
MUMPS returned INFO(1) = -9 and requires more memory, reallocating.  Attempt 3
  Increasing icntl[13] from 4000 to 8000.
____________________________________________________________________
MUMPS returned INFO(1) = -9 and requires more memory, reallocating.  Attempt 4
                                                      The time is 2024-06-26 11:09:26.879081
____________________________________________________________________
  Increasing icntl[13] from 8000 to 16000.
  17  1.1039768e+00 8.66e-01 1.04e+14  -1.7 3.14e+01    -  2.51e-01 7.10e-06h  6
____________________________________________________________________
                                                      The time is 2024-06-26 11:09:40.859015
____________________________________________________________________
  18  1.1039768e+00 8.65e-01 1.42e+12  -1.7 2.20e-04  14.0 9.86e-01 1.00e+00h  1
  19  1.1041602e+00 8.65e-01 6.82e+14  -1.7 3.20e+00    -  1.80e-02 5.74e-05h  1
____________________________________________________________________
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  20  1.1041602e+00 8.65e-01 9.67e+12  -1.7 1.07e-04  14.4 9.86e-01 1.00e+00h  1
                                                      The time is 2024-06-26 11:09:47.187771
____________________________________________________________________
In iteration 20, 1 Slack too small, adjusting variable bound
  21  1.1055720e+00 8.65e-01 5.48e+14  -1.7 3.18e+00    -  1.96e-02 4.44e-04h  1
  22r 1.1055720e+00 8.65e-01 1.00e+03  -0.1 0.00e+00    -  0.00e+00 1.21e-07R  2
____________________________________________________________________
  23r 1.2470117e+00 8.18e-01 9.95e+02  -0.1 5.49e+00    -  5.22e-03 2.58e-02f  1
                                                      The time is 2024-06-26 11:09:50.610365
  24r 1.4627598e+00 7.54e-01 9.41e+02  -0.1 6.55e+00    -  6.87e-02 3.30e-02f  1
____________________________________________________________________
  25  5.7478651e+00 1.18e-02 7.88e+04  -1.7 4.29e+00    -  3.85e-05 1.00e+00f  1
  26  4.8640796e+00 7.91e-03 7.86e+13  -1.7 2.38e+00  13.9 1.34e-01 3.71e-01h  1
____________________________________________________________________
MUMPS returned INFO(1) = -9 and requires more memory, reallocating.  Attempt 1
  Increasing icntl[13] from 16000 to 32000.
                                                      The time is 2024-06-26 11:09:55.406274
  27  4.7735730e+00 7.62e-03 9.71e+13  -1.7 2.44e+00  14.4 4.07e-01 3.70e-02h  1
____________________________________________________________________
  28  2.2891076e+00 3.26e-03 1.36e+15  -1.7 2.48e+00  13.9 4.49e-05 1.00e+00f  1
  29  3.7236717e+00 1.62e-03 6.25e+14  -1.7 4.72e+00  13.4 3.24e-01 3.04e-01h  1
____________________________________________________________________
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  30  4.5251259e+00 3.11e-04 7.44e+14  -1.7 8.01e-01  13.8 2.63e-02 1.00e+00f  1
                                                      The time is 2024-06-26 11:11:04.185239
____________________________________________________________________
  31  4.3047027e+00 2.58e-05 6.85e+13  -1.7 2.20e-01  13.4 9.37e-01 1.00e+00h  1
  32  4.3048773e+00 7.53e-10 2.66e+10  -1.7 2.55e-03  12.9 1.00e+00 1.00e+00h  1
____________________________________________________________________
  33  4.3048776e+00 2.78e-16 5.74e+06  -1.7 2.21e-06  12.4 1.00e+00 1.00e+00h  1
                                                      The time is 2024-06-26 11:11:37.837084
____________________________________________________________________
  34  4.3048776e+00 1.11e-16 5.45e-01  -1.7 6.29e-13  11.9 1.00e+00 1.00e+00h  1
  35  3.2811505e+00 5.60e-04 1.64e+00  -1.7 1.52e+00    -  7.41e-01 1.00e+00F  1
  36  3.1758015e+00 1.05e-04 1.79e-01  -1.7 5.23e-01    -  1.00e+00 1.00e+00h  1
  37  2.9511330e+00 4.33e-05 1.13e-01  -2.5 2.25e-01    -  8.82e-01 1.00e+00h  1
  38  2.5811604e+00 3.50e-04 6.39e-02  -2.5 4.76e-01    -  1.00e+00 1.00e+00h  1
  39  2.5843178e+00 8.04e-07 1.12e-03  -2.5 1.28e-01    -  1.00e+00 1.00e+00h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  40  2.2288337e+00 9.49e-05 5.70e-03  -3.8 3.55e-01    -  8.41e-01 1.00e+00h  1
  41  2.0031715e+00 5.65e-05 2.47e-02  -3.8 2.26e-01    -  1.00e+00 1.00e+00h  1
  42  1.9742973e+00 6.92e-06 1.42e-03  -3.8 1.67e-01    -  1.00e+00 1.00e+00h  1
  43  1.9318483e+00 1.45e-05 7.74e-04  -5.7 2.14e-01    -  6.20e-01 6.67e-01h  1
  44  1.9122575e+00 1.18e-05 4.29e-02  -5.7 2.79e-01    -  6.84e-01 7.73e-01h  1
  45  1.9096301e+00 6.96e-06 4.79e-02  -5.7 4.02e-01    -  6.21e-01 4.04e-01h  1
  46  1.9065888e+00 2.63e-06 1.36e-02  -5.7 4.30e-01    -  7.62e-01 7.58e-01f  1
  47  1.9055998e+00 8.93e-07 1.78e-05  -5.7 4.51e-01    -  1.00e+00 1.00e+00f  1
  48  1.9052577e+00 5.74e-07 4.31e-02  -8.6 1.57e-01    -  6.23e-01 3.75e-01h  1
  49  1.9047390e+00 2.06e-07 7.87e-03  -8.6 2.20e-01    -  7.46e-01 9.05e-01h  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
  50  1.9046844e+00 4.19e-09 3.88e-03  -8.6 2.10e-02    -  7.96e-01 9.88e-01h  1
  51  1.9046836e+00 4.65e-12 1.69e-08  -8.6 3.02e-03    -  1.00e+00 1.00e+00f  1
  52  1.9046828e+00 3.46e-13 4.47e-12  -9.0 2.17e-04    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 52

                                   (scaled)                 (unscaled)
Objective...............:   1.9046828125898771e+00    1.9046828125898771e+00
Dual infeasibility......:   4.4675374510916299e-12    4.4675374510916299e-12
Constraint violation....:   3.4627856138058632e-13    3.4627856138058632e-13
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   9.1188207858490720e-10    9.1188207858490720e-10
Overall NLP error.......:   9.1188207858490720e-10    9.1188207858490720e-10


Number of objective function evaluations             = 66
Number of objective gradient evaluations             = 51
Number of equality constraint evaluations            = 66
Number of inequality constraint evaluations          = 66
Number of equality constraint Jacobian evaluations   = 55
Number of inequality constraint Jacobian evaluations = 55
Number of Lagrangian Hessian evaluations             = 52
Total seconds in IPOPT                               = 135.225

EXIT: Optimal Solution Found.
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       nlp_f  | 644.00us (  9.76us) 235.95us (  3.58us)        66
       nlp_g  | 385.12ms (  5.84ms) 135.14ms (  2.05ms)        66
  nlp_grad_f  |   1.11ms ( 21.37us) 458.55us (  8.82us)        52
  nlp_hess_l  |   1.38 s ( 27.66ms) 634.11ms ( 12.68ms)        50
   nlp_jac_g  |   1.33 s ( 23.76ms) 571.84ms ( 10.21ms)        56
       total  | 165.46 s (165.46 s) 135.23 s (135.23 s)         1
____________________________________________________________________
                                                      The time is 2024-06-26 11:11:40.162035
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:11:41.182826
____________________________________________________________________
____________________________________________________________________
                                                      The time is 2024-06-26 11:11:42.221796
____________________________________________________________________
``` 

As you can see in the output, the PrintNowThread is blocked for long periods of time where it is unable to print. It seems this has to do with this ticket where Casadi appears to hold the GIL for longer periods of time.

Are there any plans to implement a mechanism to release the GIL when possible soon?

Thanks in advance!

**Comment by jgillis** [comment]
Hi @lfse-slafleur planned for next week..

**Comment by jgillis** [comment]
There may be python threading usecases that will fail when this feature is implemented.
Put guard on reference counting?
Overhead may be too much?

**Comment by jgillis** [comment]
Related issue: #3448

**Comment by jgillis** [comment]
```mermaid
flowchart BT

    A[WeakRefInternal] -->|raw_| B(SharedInternal)
    B -->|weak_ref_| C(WeakRef)
    C -->|node| A
    D(WeakRef) -->|node| A
    E(SharedObject) -->|node| B
    F(SharedObject) -->|node| B
  
```

**Comment by jgillis** [comment]
Remaining issue: Knitro

**Comment by jgillis** [comment]
Experiment for `det` of a `8-by-8`.
[callgrind.out (safe)](https://github.com/user-attachments/files/17629717/callgrind.out.safe.txt)
[callgrind.out (unsafe)](https://github.com/user-attachments/files/17629719/callgrind.out.unsafe.txt)

Symbolics gets ~7.5% more expensive
![Screenshot from 2024-11-05 09-55-55](https://github.com/user-attachments/assets/abd22f64-a613-4938-81e7-0283ccaea24a)

**Comment by jgillis** [comment]
`valgrind --tool=helgrind`

**Comment by jgillis** [comment]
Finished. To be merged in 3.7

**Comment by jgillis** [comment]
Idea: make plugin calls have a mutex by default; we don't want third-part plugins to randomly crash on experimenting users.
